[BUG] client timeout when invoking and agent referencing other agents

### 📋 Prerequisites

- [x] I have searched the [existing issues](./issues) to avoid creating a duplicate
- [x] By submitting this issue, you agree to follow our [Code of Conduct](https://github.com/kagent-dev/kagent/blob/main/CODE_OF_CONDUCT.md)
- [x] I am using the latest version of the software
- [x] I have tried to clear cache/cookies or used incognito mode (if ui-related)
- [x] I can consistently reproduce this issue

### 🎯 Affected Service(s)

Multiple services / System-wide issue

### 🚦 Impact/Severity

No impact (Default)

### 🐛 Bug Description

When an agent references other agents, the call times out.

Controller:

```
2025-09-12T19:00:45Z    ERROR   client/client.go:259    Error reading SSE stream for request : context deadline exceeded (Client.Timeout or context cancellation while reading body)
```

### 🔄 Steps To Reproduce

1. Create a new agent
2. Add other agents to it
3. invoke it

### 🤔 Expected Behavior

_No response_

### 📱 Actual Behavior

_No response_

### 💻 Environment

_No response_

### 🔧 CLI Bug Report

_No response_

### 🔍 Additional Context

_No response_

### 📋 Logs

```shell

```

### 📷 Screenshots

_No response_

### 🙋 Are you willing to contribute?

- [ ] I am willing to submit a PR to fix this issue

--- a/helm/kagent/values.yaml
+++ b/helm/kagent/values.yaml
@@ -72,7 +72,7 @@ controller:
   streaming: # Streaming buffer size for A2A communication
     maxBufSize: 1Mi # 1024 * 1024
     initialBufSize: 4Ki # 4 * 1024
-    timeout: 60s # 60 seconds
+    timeout: 600s # 600 seconds
   # -- Namespaces the controller should watch.
   # If empty, the controller will watch ALL available namespaces.
   # @default -- [] (watches all available namespaces)

--- a/ui/src/app/a2a/[namespace]/[agentName]/route.ts
+++ b/ui/src/app/a2a/[namespace]/[agentName]/route.ts
@@ -49,20 +49,39 @@ export async function POST(
       'Access-Control-Allow-Headers': 'Content-Type, Authorization, Accept',
     });
 
+    const KEEP_ALIVE_INTERVAL_MS = 30000; // 30 seconds
     const stream = new ReadableStream({
       start(controller) {
         const reader = backendResponse.body!.getReader();
         const decoder = new TextDecoder();
         let buffer = "";
         let isClosed = false;
+        let keepAliveTimer: NodeJS.Timeout | null = null;
+
+        // Send keep-alive comment event and reschedule timer
+        const sendKeepAlive = () => {
+          if (!isClosed) {
+            controller.enqueue(new TextEncoder().encode(': keep-alive\n\n'));
+            resetKeepAliveTimer(); // Reschedule after sending keep-alive
+          }
+        };
+
+        // Keep-alive timer logic
+        const resetKeepAliveTimer = () => {
+          if (keepAliveTimer) clearTimeout(keepAliveTimer);
+          keepAliveTimer = setTimeout(sendKeepAlive, KEEP_ALIVE_INTERVAL_MS);
+        };
 
         const pump = (): Promise<void> => {
           return reader.read().then(({ done, value }): Promise<void> => {
             if (done) {
               if (!isClosed) {
+                if (keepAliveTimer) clearTimeout(keepAliveTimer);
                 controller.close();
                 isClosed = true;
+                console.log('A2A Proxy: Backend connection closed.');
               }
+
               return Promise.resolve();
             }
 
@@ -78,16 +97,21 @@ export async function POST(
                 const eventData = eventText + '\n\n';
                 if (!isClosed) {
                   controller.enqueue(new TextEncoder().encode(eventData));
+                  resetKeepAliveTimer(); // Reset keep-alive timer whenever data is sent
                 }
               }
             }
 
             return pump();
           }).catch(error => {
+            console.error('A2A Proxy: Error in stream pump:', error);
+            if (keepAliveTimer) clearTimeout(keepAliveTimer);
+            
             if (!isClosed) {
               controller.error(error);
               isClosed = true;
             }
+
             return Promise.resolve();
           });
         };

--- a/ui/src/components/chat/ChatInterface.tsx
+++ b/ui/src/components/chat/ChatInterface.tsx
@@ -247,7 +247,7 @@ export default function ChatInterface({ selectedAgentName, selectedNamespace, se
           }
         };
 
-        // Start timeout timer 
+        // Start timeout timer
         const startTimeout = () => {
           if (timeoutTimer) clearTimeout(timeoutTimer);
           timeoutTimer = setTimeout(handleTimeout, streamTimeout);

--- a/ui/src/components/chat/ChatInterface.tsx
+++ b/ui/src/components/chat/ChatInterface.tsx
@@ -225,32 +225,57 @@ export default function ChatInterface({ selectedAgentName, selectedNamespace, se
           message: a2aMessage,
           metadata: {}
         };
-        const stream = await kagentA2AClient.sendMessageStream(selectedNamespace, selectedAgentName, sendParams);
+        const stream = await kagentA2AClient.sendMessageStream(
+          selectedNamespace,
+          selectedAgentName,
+          sendParams,
+          abortControllerRef.current?.signal
+        );
 
         let lastEventTime = Date.now();
-        const streamTimeout = 60000;
-
-        for await (const event of stream) {
-          lastEventTime = Date.now();
-
-          try {
-            handleMessageEvent(event);
-          } catch (error) {
-            console.error("❌ Event that caused error:", event);
+        let timeoutTimer: NodeJS.Timeout | null = null;
+        let streamActive = true;
+        const streamTimeout = 600000; // 10 minutes
+        
+        // Timeout handler
+        const handleTimeout = () => {
+          if (streamActive) {
+            console.error("⏰ Stream timeout - no events received for 10 minutes");
+            toast.error("⏰ Stream timed out - no events received for 10 minutes");
+            streamActive = false;
+            if (abortControllerRef.current) abortControllerRef.current.abort();
           }
+        };
 
-          // Check if we should stop streaming due to cancellation
-          if (abortControllerRef.current?.signal.aborted) {
-            break;
-          }
+        // Start timeout timer 
+        const startTimeout = () => {
+          if (timeoutTimer) clearTimeout(timeoutTimer);
+          timeoutTimer = setTimeout(handleTimeout, streamTimeout);
+        };
+        startTimeout();
 
-          // Timeout check (in case stream hangs)
-          if (Date.now() - lastEventTime > streamTimeout) {
-            console.warn("⏰ Stream timeout - no events received for 30 seconds");
-            break;
+        try {
+          for await (const event of stream) {
+            lastEventTime = Date.now();
+            startTimeout(); // Reset timeout after every event
+
+            try {
+              handleMessageEvent(event);
+            } catch (error) {
+              console.error("❌ Event that caused error:", event);
+            }
+
+            // Check if we should stop streaming due to cancellation
+            if (abortControllerRef.current?.signal.aborted) {
+              console.info("Stream aborted");
+              streamActive = false;
+              break;
+            }
           }
+        } finally {
+          streamActive = false;
+          if (timeoutTimer) clearTimeout(timeoutTimer);
         }
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
       } catch (error: any) {
         if (error.name === "AbortError") {
           toast.info("Request cancelled");

--- a/ui/src/lib/a2aClient.ts
+++ b/ui/src/lib/a2aClient.ts
@@ -38,15 +38,17 @@ export class KagentA2AClient {
 
   /**
    * Send a streaming message using the A2A protocol via Next.js API route
+   * Accepts an optional AbortSignal for cancellation support
    */
   async sendMessageStream(
-    namespace: string, 
-    agentName: string, 
-    params: MessageSendParams
+    namespace: string,
+    agentName: string,
+    params: MessageSendParams,
+    signal?: AbortSignal
   ): Promise<AsyncIterable<any>> {
     const request = this.createStreamingRequest(params);
-    // This redirects to the Next.js API route 
-    // Note that this route CAN'T be the same 
+    // This redirects to the Next.js API route
+    // Note that this route CAN'T be the same
     // as the routes on the backend.
     const proxyUrl = `/a2a/${namespace}/${agentName}`;
 
@@ -57,6 +59,7 @@ export class KagentA2AClient {
         'Accept': 'text/event-stream',
       },
       body: JSON.stringify(request),
+      signal,
     });
 
     if (!response.ok) {
@@ -102,11 +105,11 @@ export class KagentA2AClient {
             for (const line of lines) {
               if (line.startsWith('data: ')) {
                 const dataString = line.substring(6);
-                
+
                 if (dataString === '[DONE]') {
                   return;
                 }
-                
+
                 try {
                   const eventData = JSON.parse(dataString);
                   yield eventData.result || eventData;

